components/addressmanager/src/lib.rs

```diff
@@ -120,76 +107,89 @@ mod not_banned_address_store_with_cache {
-            self.db_store.set(address.ip, address.port, connection_failed_count).unwrap();
-            self.addresses.insert(NetAddress::new(address.ip, address.port), connection_failed_count);
+        pub fn set(&mut self, address: NetAddress, connection_failed_count: u64) {
+            let entry = match self.addresses.get(&address.into()) {
+                Some(entry) => Entry { connection_failed_count, address: entry.address },
+                None => Entry { connection_failed_count, address },
+            };
+            self.db_store.set(address.into(), entry).unwrap();
+            self.addresses.insert(address.into(), entry);
             self.keep_limit();
         }
 
         fn keep_limit(&mut self) {
             while self.addresses.len() > MAX_ADDRESSES {
-                let to_remove = self.addresses.iter().max_by(|a, b| (a.1).0.cmp(&(b.1).0)).unwrap();
-                self.remove(*to_remove.0);
+                let to_remove =
+                    self.addresses.iter().max_by(|a, b| (a.1).connection_failed_count.cmp(&(b.1).connection_failed_count)).unwrap();
+                self.remove_by_key(*to_remove.0);
             }
         }
 
-        pub fn get(&self, address: NetAddress) -> ConnectionFailureCount {
-            *self.addresses.get(&address).unwrap()
+        pub fn get(&self, address: NetAddress) -> Entry {
+            *self.addresses.get(&address.into()).unwrap()
         }
 
         pub fn remove(&mut self, address: NetAddress) {
-            self.addresses.remove(&address);
-            self.db_store.remove(address.ip, address.port).unwrap()
+            self.remove_by_key(address.into())
+        }
+
+        fn remove_by_key(&mut self, key: AddressKey) {
+            self.addresses.remove(&key);
+            self.db_store.remove(key).unwrap()
         }
 
         pub fn get_all_addresses(&self) -> impl Iterator<Item = NetAddress> + '_ {
-            self.addresses.keys().copied()
+            self.addresses.values().map(|entry| entry.address)
         }
 
-        pub fn get_random_addresses(&self, count: usize, exceptions: HashSet<NetAddress>) -> Vec<NetAddress> {
+        pub fn get_randomized_addresses(&self, exceptions: HashSet<NetAddress>) -> Vec<NetAddress> {
+            let exceptions: HashSet<AddressKey> = exceptions.into_iter().map(|addr| addr.into()).collect();
             let addresses = self.addresses.iter().filter(|(addr, _)| !exceptions.contains(addr)).collect_vec();
-            let count = min(count, addresses.len());
-            let mut weights =
-                addresses.iter().map(|(_, count)| 64f64.powf((MAX_CONNECTION_FAILED_COUNT + 1 - count.0) as f64)).collect_vec();
+            let mut weights = addresses
+                .iter()
+                .map(|(_, entry)| 64f64.powf((MAX_CONNECTION_FAILED_COUNT + 1 - entry.connection_failed_count) as f64))
+                .collect_vec();
 
-            (0..count)
+            (0..addresses.len())
```

Comment:
If we go through all addresses (and eventually select all), it means we are only randomly sorting with priority, I would expect much more efficient methods. This imp makes sense to me only with `count` << `addresses.len()`